Scans

michaelcfanning opened the pull request.

eddynaka commented on src/Microsoft.Security.Utilities/IdentifiableSecrets.cs:
```diff
@@ -16,7 +19,7 @@ public static class IdentifiableSecrets
-        public static string GenerateIdentifiableKey(ulong checksumSeed,
+        public static string GenerateBase64Key(ulong checksumSeed,
                                                      uint keyLengthInBytes,
                                                      string base64EncodedSignature)
```
pls, fix the indentation

eddynaka commented:
Change looks good.
Can you also update the title of the PR, pls?

michaelcfanning reviewed (approved):
:shipit:

michaelcfanning commented:
We forgot the release notes update on this one. Every method name change is a breaking change for customers. I will submit a  new PR with this, stay tuned.

---
In reply to: [1013659836](https://github.com/microsoft/security-utilities/pull/10#issuecomment-1013659836) [](http://example.com/codeflow?ancestors=1013659836,1013659836)

beaubelgrave committed:
Rust/His: Add fast byte scanner for HISv2

We need a way to quickly scan for highly identifiable secrets in both
files and raw byte streams of unknown encodings. Due to the way HISv2
works, we can scan for the signature JQQJ and know if the data is UTF8
or UTF16.

To enable this quickly, we first scan 16 bytes at a time for the letter
J via loop that LLVM vectorizes. If the letter J is found, then we start
acccumulating bytes into a 64-bit value. 8 bytes is enough to detect
JQQJ in UTF8 and UTF16. We can check all bytes in a single instruction
vs having to check one by one. We only have to compare the accumulator
once we add J to the accumulator.

Once we detect JQQJ, we save the location the detection for later
further full analysis. This allows for the scanner to scan large files
and only save the possible locations to check. It can also scan small or
large byte buffers.

The possible matches are fetched from memory or a file, then they are
sent through a full validation of the data. The full validation so far
is only the pattern/regex matching. It does not currently include the
marvin checksum, etc.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
Rust/His: Fix typo and add file unit tests

There was a typo in the matching of an equals sign, which this fixes.

Add unit tests for file readings, which uncovered a corner case. We
cannot assume explicit sizes of bytes, since files may be truncated. We
must move to a range of bytes when determining how to parse the bytes.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
Rust/His: Add HisV1 32 and 64 byte signatures

We have various HisV1 4 byte signatures, each indicates the byte length
of the keys. Many of these are 32 and 64 byte signatures.

Add initial support for scanning for only v2, v1, or both.
Add unit test to ensure known signatures match for 32 and 64 byte V1.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
His/Rust: Fix bug and ensure ScanMatchType is ABI

ScanMatchType will be exposed out via FFI to other languages. We need to
ensure these values do not change.

Define constants to ScanMatchType to ensure they don't change.
Ensure we always check byte range before checking patterns.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
His/Rust: Expose Scan and ScanOptions for FFI

Scan and ScanOptions are behind a non-public module.

Ensure we have both Scan and ScanOptions exported for callers to
utilize.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
His/Rust: Make FFI compatible crate

Add an FFI compatible crate that can be called from C like languages.

This crate just links in the normal core security crate and wraps it up
with C exports.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

beaubelgrave committed:
His/DotNet: Utilize the FFI core crate for scanning

Add a new IdentifiableScan class in the core DLL that utilizes the FFI
crate dylib.

Update the CLI tool to have a new "scan" command that uses the new
IdentifiableScan class to scan for identifiable secrets within a given
file.

Signed-off-by: Beau Belgrave <beaub@linux.microsoft.com>

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+
+/*
+ * NOTE:
+ * These are exposed for FFI, we must keep the numbers consistent.
```
What is FFI?

beaubelgrave replied:
Foreign Function Interface:
https://doc.rust-lang.org/nomicon/ffi.html

suvamM replied:
@beaubelgrave Got my answer: foreign function interface

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+    His32Utf16 = 2,
+    His64Utf8 = 3,
+    His64Utf16 = 4,
+    His2Utf8 = 5,
```
The naming might be confusing to a new reader: `His2` is referring to a version whereas `His32` isn't. Perhaps we can make this explicit via `His_32...` and `His2_...`?

beaubelgrave replied:
Yeah, I agree. Rust / clippy wanted these to be named a certain way to avoid a warning.

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+
+        /* NOTE: We skip the signature since we already checked */
+
+        if data[37] < b'A' || data[37] > b'P' {
```
For performance reasons, would it make sense to perform this check first? In other words, perform checks in ranked in order of byte array length?

suvamM replied:
@beaubelgrave I think I see what's happening here: you have a fast check for the signature in `match_sig_v1` that precisely checks for these signatures. Then the order of the checks in this method does not matter much. Is that correct? 
Same logic for my remaining comments on the ordering of the conditions.

beaubelgrave replied:
Yeah, this part is the "slow" path and doesn't need to be fast. It already requires another data offset slice (or file read) after the initial scan targets the 4-byte signature.

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+            }
+        }
+
+        43
```
Can we use a variable instead of a hard coded constant?

suvamM replied:
`HIS_32_UTF8_LEN - 1`?

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+            }
+        }
+
+        if data[85] != b'A' && data[85] != b'Q' &&
```
Same feedback here: perhaps it might be worth moving this check to the start, and performing the remaining checks in order of byte array length?

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+            return HIS_64_UTF8_LEN;
+        }
+
+        86
```
`HIS_64_UTF8_LEN - 2`?

suvamM commented on src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_scans.rs:
```diff
@@ -0,0 +1,1116 @@
+            }
+        }
+
+        if &data[52..58] != b"JQQJ99" {
```
If the byte representation of `JQQJ99` in the appropriate location, then we shouldn't even look at the remaining bytes.
